### PR TITLE
Raise warning for unpickable local function

### LINKED ISSUE #538: Warn about pickle-ablity when using `dp.map(some_local_function)` ?
`torchdata` issues a warning about pickle when we use lambdas (which is great!)
Another kind of function that isn't compatible with pickle are local functions. Would it be possible to throw the same warning there?

```py
import torchdata
import pickle

def make_dp():

    def f(x):  # local function, not pickleable
        return x

    return torchdata.datapipes.iter.IterableWrapper(range(40)).map(f)

dp = make_dp()  # no warning

f = "/tmp/dp"
pickle.dump(dp, open(f, "wb"))  # fails
```

### PR DESCRIPTION
**Pull request by ejguan**
Fixes https://github.com/pytorch/data/issues/538

- Improve the validation function to raise warning about unpickable function when either lambda or local function is provided to `DataPipe`.
- The inner function from `functools.partial` object is extracted as well for validation
- Mimic the behavior of `pickle` module for local lambda function: It would only raise Error for the local function rather than `lambda` function. So, we will raise warning about local function not lambda function.
```py
>>> import pickle
>>> def fn():
...     lf = lambda x: x
...     pickle.dumps(lf)
>>> pickle.dumps(fn)
AttributeError: Can't pickle local object 'fn.<locals>.<lambda>'
```

### CODE REVIEW

**Review by NicolasHug — COMMENTED** [review]
Thanks for the quick fix!

File: torch/utils/data/datapipes/utils/common.py
Comment (ejguan) [reply torch/utils/data/datapipes/utils/common.py:27]: Wow. Let me take a look.

File: test/test_datapipe.py
Comment (VitalyFedyunin) [inline test/test_datapipe.py:758]: super nit: consider using one of `assertWarns` `assertWarnsRegexp` https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertWarnsRegex

**Review by VitalyFedyunin — APPROVED** [review]

**Review by NivekT — APPROVED** [review]

**Comment by ejguan** [comment]
I might wait until #79344 landed

**Comment by ejguan** [comment]
Closing the PR to land this one from internal phabricator

**Pull request closed without merging**
